app/viewmodels/helpers/TimelineHelper.scala

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright 2023 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package viewmodels.helpers
+
+import models.response.emcsTfe.getMovementHistoryEvents.{GetMovementHistoryEventsResponse, MovementHistoryEvent}
+import play.api.i18n.Messages
+import utils.{DateUtils, Logging}
+import viewmodels.TimelineEvent
+
+import java.time.format.DateTimeFormatter
+import java.time.{LocalDate, LocalDateTime}
+import javax.inject.Inject
+import scala.util.Try
+
+class TimelineHelper @Inject()() extends Logging with DateUtils {
+
+  def timeline(historyEventsResponse: GetMovementHistoryEventsResponse)(implicit messages: Messages): Seq[TimelineEvent] =
+    historyEventsResponse.movementHistory.map { event =>
+      TimelineEvent(
+        eventType = event.eventType,
+        title = messages(getEventTitleKey(event)),
+        date = parseDateTime(event.eventDate),
+        url = getHistoryEventUrl(event)
+      )
+    }.sortBy(_.date)
+
+
+  def getEventTitleKey(event: MovementHistoryEvent): String = {
+    (event.eventType, event.sequenceNumber, event.messageRole) match {
+      case ("IE801", 1, 0) => s"movementHistoryEvent.${event.eventType}.first.label"
+      case ("IE801", 2, 0) => s"movementHistoryEvent.${event.eventType}.further.label"
+      case ("IE802", 0, 1) => s"movementHistoryEvent.${event.eventType}.cod.label"
+      case ("IE802", 0, 2) => s"movementHistoryEvent.${event.eventType}.ror.label"
+      case ("IE802", 0, 3) => s"movementHistoryEvent.${event.eventType}.des.label"
+      case ("IE803", 0, 1) => s"movementHistoryEvent.${event.eventType}.diverted.label"
+      case ("IE803", 0, 2) => s"movementHistoryEvent.${event.eventType}.split.label"
+      case ("IE840", 0, 1) => s"movementHistoryEvent.${event.eventType}.first.label"
+      case ("IE840", 0, 2) => s"movementHistoryEvent.${event.eventType}.complementary.label"
+      case _ => s"movementHistoryEvent.${event.eventType}.label"
+    }
+
+  }
+
+  def parseDateTime(eventDate: String): LocalDateTime =
+    Try {
+      LocalDateTime.parse(eventDate, DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss"))
+    }.getOrElse{
+      logger.error(s"[parseDateTime] - un-parseable date/time received [$eventDate]")
+      LocalDate.now.atStartOfDay
+    }
+
+  def getHistoryEventUrl(event: MovementHistoryEvent): String =
+    s"event/${event.eventType}/id/${event.eventDate.hashCode >>> 1}"
```

Comment:
Never seen three `>>>` used before. What does it do out of interest?

Reply:
Thats a bit shift to the right of 1. The hashcode of the date/time combo could return a negative number so the bit shift to the right ensures its a positive number (as this appears in the URL)

Reply:
https://www.programiz.com/java-programming/bitwise-operators#unsigned-right-shift